pkg/controllers/node/node_controller.go

```diff
@@ -0,0 +1,283 @@
+func getK8sNodeRef(calicoNode api.Node) (string, error) {
+	for _, orchRef := range calicoNode.Spec.OrchRefs {
+		if orchRef.Orchestrator == "k8s" {
+			return orchRef.NodeName, nil
+		}
+	}
+	return "", fmt.Errorf("Node does not have a kubernetes nodeRef")
+}
+
+// Run starts the node controller. It does start-of-day preparation
+// and then launches worker threads.
+func (c *NodeController) Run(threadiness int, reconcilerPeriod string, stopCh chan struct{}) {
+	defer uruntime.HandleCrash()
+
+	// Let the workers stop when we are done
+	workqueue := c.nodeDataCache.GetQueue()
+	defer workqueue.ShutDown()
+
+	log.Info("Starting Node controller")
+
+	// Load node cache. Retry when failed.
+	log.Debug("Loading node cache at start of day")
+	for err := c.populateCache(); err != nil; {
+		log.WithError(err).Errorf("Failed to load Node cache, retrying in 5s")
+		time.Sleep(5 * time.Second)
+	}
+
+	// Wait till k8s cache is synced
+	go c.informer.Run(stopCh)
+	log.Debug("Waiting to sync with Kubernetes API (Nodes)")
+	for !c.informer.HasSynced() {
+	}
+	log.Debug("Finished syncing with Kubernetes API (Nodes)")
+
+	// Start Calico cache.
+	c.nodeDataCache.Run(reconcilerPeriod)
+
+	// Start a number of worker threads to read from the queue.
+	for i := 0; i < threadiness; i++ {
+		go c.runWorker()
+	}
+	log.Info("Node controller is now running")
+
+	<-stopCh
+	log.Info("Stopping Node controller")
+}
+
+func (c *NodeController) runWorker() {
+	for c.processNextItem() {}
+}
+
+func (c *NodeController) processNextItem() bool {
+	// Wait until there is a new item in the work queue.
+	workqueue := c.nodeDataCache.GetQueue()
+	key, quit := workqueue.Get()
+	if quit {
+		return false
+	}
+
+	// Sync the object to the Calico datastore.
+	if err := c.syncToCalico(key.(string)); err != nil {
+		c.handleErr(err, key.(string))
+	}
+
+	// Indicate that we're done processing this key, allowing for safe parallel processing such that
+	// two objects with the same key are never processed in parallel.
+	workqueue.Done(key)
+	return true
+}
+
+// populateCache fills the nodeCache with initial data
+// by querying the existing data stored in etcd.
+func (c *NodeController) populateCache() error {
+	nodes, err := c.calicoClient.Nodes().List(api.NodeMetadata{})
+	if err != nil {
+		return err
+	}
+
+	c.nodeCache.Lock()
+	for _, node := range nodes.Items {
+		k8sNodeName, err := getK8sNodeRef(node)
+		if err != nil {
+			c.nodeCache.data[k8sNodeName] = node.Metadata.Name
+		}
+
+	}
+	log.Infof("Current Cache: %v", c.nodeCache.data)
+	c.nodeCache.Unlock()
+	return nil
+}
+
+// syncToCalico syncs the given update to the Calico datastore.
+func (c *NodeController) syncToCalico(key string) error {
+	// Check if it exists in the controller's cache.
+	_, exists := c.nodeDataCache.Get(key)
+	log.Infof("%v", key)
+	if !exists {
+		// The object no longer exists - delete from the datastore.
+		log.Infof("Deleting node %s from Calico datastore", key)
```

Comment:
We should move this log a bit lower down, to where we actually are about to send the Delete request to libcalico.

At this point in execution, we don't know if we're actually going to delete it from the datastore.

We could / should still have a debug log here which says "Node %s no longer exists"